src/routes/utils/user.ts

```diff
@@ -0,0 +1,102 @@
+import { FastifyInstance, FastifyReply, FastifyRequest } from 'fastify';
+import { Type, Static } from '@sinclair/typebox';
+import { knexController } from '../../database/database';
+
+import bcrypt from 'bcrypt';
+
+const editUserResponse = Type.Object({
+  email: Type.String(),
+  forename: Type.String(),
+  surname: Type.String(),
+  password: Type.String(),
+});
+
+export type EditUserResponse = Static<typeof editUserResponse>;
+
+// Expected request type
+interface User {
+  email: string;
+  forename: string;
+  surname: string;
+  password: string;
+}
+
+const editUserSchema = {
+  description: 'Edit data of already existing user.',
+  summary: 'Edit user',
+  tags: ['User'],
+  body: {
+    type: 'object',
+    required: ['email', 'password', 'forename', 'surname'],
+    properties: {
+      email: {
+        type: 'string',
+      },
+      forename: {
+        type: 'string',
+      },
+      surname: {
+        type: 'string',
+      },
+      password: {
+        type: 'string',
+      },
+    },
+  },
+  response: {
+    200: editUserResponse,
+  },
+};
+
+const hashPassword = async (
+  password: string
+): Promise<{ hash: string; salt: string }> => {
+  const saltRounds = 10;
+
+  const salt = bcrypt.genSaltSync(saltRounds);
+  const hash = bcrypt.hashSync(password, salt);
+
+  return { hash, salt };
+};
+
+const editUserHandler = async (
+  req: FastifyRequest<{ Body: User }>,
+  reply: FastifyReply
+): Promise<void> => {
+  const { email, password, forename, surname } = req.body;
+
+  // hash password.
+  const hashObj = await hashPassword(password);
+
+  // TODO check if passwords match before editing.
+
+  // edit user (TODO edit email?)
+  const editResponse = await knexController('user').where({ email }).update({
+    password_hash: hashObj.hash,
+    forename,
+    surname,
+    salt: hashObj.salt,
+  });
+
+  // If no user found with given email.
+  if (editResponse === 0) {
+    throw new Error('No user found with given email.');
+  }
+
+  // get edited user
+  const editedUser = await knexController
+    .select('email', 'password_hash as password', 'forename', 'surname')
+    .from<User>('user')
+    .where({ email });
```

Comment:
Jos rivillä 74 antaa .update'lle toiseksi parametriksi listan ['email', 'forename', 'surname'], niin se palauttaa objektin jossa on kyseiset sarakkeet. Salasanan hashiä ei varmaankaan kannata palauttaa.

Reply:
Onko varma? Koitin sitä saada toimimaan ja palautti aina vain muutettujen määrän.  Voi olla että tein jotain väärin tietty.

Koskien ainakin MySql :
https://stackoverflow.com/questions/41598701/how-to-get-a-list-of-all-updated-records-in-knex-mysql

Reply:
https://knexjs.org/guide/query-builder.html#update mukaan pitäis palauttaa
Disclaimer: en oo ite kokeillu :D